picard/util/pipe.py

```diff
@@ -0,0 +1,430 @@
+
+    def __str__(self) -> str:
+        messages_str = "\n  ".join(str(m) for m in self.messages)
+        if not messages_str:
+            messages_str = "unknown"
+        return f"ERROR: {messages_str}"
+
+
+class PipeErrorInvalidArgs(PipeError):
+    MESSAGE = "Pipe() args argument has to be iterable"
+
+
+class PipeErrorInvalidAppData(PipeError):
+    MESSAGE = "Pipe() app_name and app_version arguments have to be str"
+
+
+class PipeErrorNotFound(PipeError):
+    MESSAGE = "Pipe doesn't exist"
+
+
+class PipeErrorBroken(PipeError):
+    MESSAGE = "Pipe is broken"
+
+
+class PipeErrorInvalidResponse(PipeError):
+    MESSAGE = "Invalid response from pipe"
+
+
+class PipeErrorWin(PipeError):
+    MESSAGE = "Windows API error"
+
+
+class PipeErrorNoPermission(PipeError):
+    MESSAGE = "No permissions for creating a pipe"
+
+
+class PipeErrorNoDestination(PipeError):
+    MESSAGE = "No available dirs to place a pipe"
+
+
+class AbstractPipe(metaclass=ABCMeta):
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    MESSAGE_TO_IGNORE: str = '\0'
+    TIMEOUT_SECS: float = 1.5
+
+    @classmethod
+    @property
+    @abstractmethod
+    def PIPE_DIRS(cls):
+        """
+        Tuple of dirs where pipe could possibly be created
+
+        **Virtual**, implement in child classes
+        """
+        raise NotImplementedError
+
+    def __init__(self, app_name: str, app_version: str, args: Optional[Iterable[str]] = None,
+                 forced_path: Optional[str] = None):
+        """
+        :param app_name: (str) Name of the app, included in the pipe name
+        :param app_version: (str) Version of the app, included in the pipe name
+        :param args: (Optional[Iterable[str]]) Will be passed to an existing instance of app if possible
+        :param forced_path: (Optional[str]) Testing-purposes only, bypass of no $HOME on testing machines
+        """
+        if args is None:
+            self._args: Tuple[str] = tuple()    # type: ignore
+        else:
+            try:
+                self._args = tuple(args)    # type: ignore
+            except TypeError as exc:
+                raise PipeErrorInvalidArgs(exc) from None
+
+        if not self._args:
+            self._args = (self.MESSAGE_TO_IGNORE,)
+
+        if not isinstance(app_name, str) or not isinstance(app_version, str):
+            raise PipeErrorInvalidAppData
+
+        if forced_path:
+            self._paths = (forced_path,)
+        elif IS_WIN or os.getenv("HOME"):
+            self._paths = self.__generate_filenames(app_name, app_version)
+            self.path_was_forced = False
+        else:
+            self._paths = (NamedTemporaryFile(delete=False).name,)
+            self.path_was_forced = True
+            log.debug("Pipe path had to be mocked by a temporary file")
+        self.is_pipe_owner: bool = False
+
+        self.__thread_pool = concurrent.futures.ThreadPoolExecutor()
```

Comment:
The default `max_workers` value is likely much higher than our needs, it might be worth specifying it.

https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor

Reply:
I'm thinking about 6 of them. 2 per reader, 2 per sender (they need two workers for the sake of *hacky killing the job*) and 2 just in case.

1 more than default value for 1 CPU, 28 less in the edge case